### PR TITLE
[KS] ProgressMeter Feature SVG Element with transparancy and responsibility

### PR DESCRIPTION
**Pull request by qdittrich**
This feature for the ProgressMeter KitchenSink component is step two of the bugfix- / feature-bundle, mentioned in the ticket below.
This feature has effect on the standard and fixed-size version of the component. While developers may not need to change the function calls, the visualization slightly changes, so now the background is transparent and the element is responible. Also the fixed-size version now supports a fifth value to directly set the size.

https://mantis.ilias.de/view.php?id=23654

### CODE REVIEW

**Comment by qdittrich** [comment]
The PHPUnit tests are not edited yet. I update this PR when it is done

**Comment by qdittrich** [comment]
The PHPUnit tests are updated and the code got a cleanup. From my view, it's ready to get merged.

File: src/UI/Implementation/Component/Chart/ProgressMeter/Renderer.php
Comment (klees) [inline src/UI/Implementation/Component/Chart/ProgressMeter/Renderer.php]: Change Request: Please just remove this code if we do not need it.

**Review by klees — CHANGES_REQUESTED** [review]
Hi @qdittrich!

Thanks a lot for this contribution to ILIAS.

Please consider the following suggestions. You do not need to follow those, but but please indicate shortly
why you prefer to do otherwise:
- [ ] Deduplicate code in the Renderer. The code in the lines 63 to 98 and lines 135 to 171 looks very similar.

Please implement the following changes:
- [ ] Remove uncommented code that we do not need anymore (see Renderer.php, l. 238)
- [ ] Remove the "width"-Parameter on the `FixedSize`-class (and the related code as well). The sizing of UI-components should only depend on the style-sheets and not be determined by developers implementing stuff using the UI-components. On the one hand, the developer can not sensibly determine the required size, as it depends on stuff that is not under her control. On the other hand, setting a style directly on an element does not allow to change that style via the style sheet.

It is possible that @Amstutz will have questions, suggestions or change requests as well, so you might want to wait what he has to add before you return to coding. Thanks!

Best regards!
@klees

**Comment by qdittrich** [comment]
Hi @klees ,

Thank you for your reply.

You are absolutely right about the commented out lines. I just overlooked them at the cleanup step.

The doubled code between the functions comes from the fact, that I tried to get the changes as small as possible. But I think you are right that the prevention of code doubling is more important than the amount of changes. I will have a look at this again.

About the "width" parameter it is a little bit tricky. As this component was introduced, there was implemented a responsive and a fixed-size version. Now that the components output was changed from html to svg, the fixed-size version is kind of special in some way. I prefered to use the size given at the moment of instanciating the component to get the possibility of different sizes in different places. Maybe it would be a better way to give something like a special identifier (data-parameter or something) to size it by stylesheets. I'm not sure, at the moment, what would be the best way.
If you have an good idea how I should handle this, feel free contact me :-)

Best regards!
@qdittrich

**Comment by klees** [comment]
Hi @qdittrich!

Thanks for your changes so far!

About the sizing of the icon: If we'd really need different sizes for the fixed size progressmeter, we could approach that similar to `Icon` and just name the sizes (e.g. small, medium or large) and then assign an according class to the element in html. An alternative would be to just introduce new constructors instead of fixedSize, e.g. `midi` or `maxi`, which, given the current `mini`-constructor, is what I would prefer in this case. `standard` that would need to be named `responsive`, right?

I still would not introduce that without a usecase. It is perfectly possible to set the sizes for the progressmeter via CSS atm, even in different contexts if required. I currently cannot imagine a case where we would require developers to think about the sizes of the progressmeter. Do you see such a case?

Best regards!

**Comment by qdittrich** [comment]
Hi @klees ,

thank you for your suggestion!

I think to rename the constructors would get a little to far away from the original introduction. The ProgressMeter was accepted by the JF with the three given constructors and if I remember correctly, such a change of the constructors would need a new JF desicion.

But you are absolutely right by mentioning that the developer should not need to think about display sizes. In view of that, I will remove the width parameter for the moment and reinsert the class "il-chart-progressmeter-fixed-size". So the behavior of the fixed-size version will be like before this feature.

What do you think?

In the future I could think about adding different classes (maybe given in the constructor) but for now one class should be enough. And like you mentioned previously, I do not know if there will be any case where multiple "fixed-size" sizes are wanted or needed.


Because you mentioned the "mini" version: This is no longer mentioned to be only a small graphic. Instead it is more like a minimalistic or minified version. I will also introduce a further PR for the change of the description. Because this needs a JF desicion, I extracted that change and want to start the PR after the this onw is merged. This is just for your information.

Best regards!

**Comment by klees** [comment]
Hi @qdittrich, sounds plausibel. Thx a lot!

**Review by Amstutz — APPROVED** [review]
Hi @qdittrich!

Thanks a lot for this contribution to ILIAS.

I approve the suggested changes. Note that they include a change of interface and must therefore be accepted by the JF. Since you will not be able to attend, I will make the case and suggest to accept your changes there. I will merge this afterwards if no objections are voiced.

Best regards!
@Amstutz 

[Edit]
The tests are currently failing, this needs to be fixed before we can merge.

Please implement the following changes:
- [ ] Fix Failing Unit Tests

[Edit]

**Comment by qdittrich** [comment]
Hi @Amstutz I fixed the unit tests for the ProgressMeter.
The tests fill fail though, because the `CrawlerTest::testParseProperEntryToYamlEntries` seems to fails on another component.

**Comment by qdittrich** [comment]
@Amstutz  I just saw that you mentioned a change in the component description, that needs to be approved by the JF. The content was changed earlier but was reverted. The only change at the description now, is a added space to correct the indentation of one line. Does this really needs a JF decision?

File: src/UI/Implementation/Component/Chart/ProgressMeter/Renderer.php
Comment (qdittrich) [reply src/UI/Implementation/Component/Chart/ProgressMeter/Renderer.php]: @klees Could you check your change requests (also in the comment above) and mark them as solved? If I have overlooked something, please mention me again.

**Comment by Amstutz** [comment]
@qdittrich I am sry, I missed that. Corrected Intention obviously doe not need JF approval ;-) So my points are all taken care of. @klees are you fine with merging, as well? If so, click the button or ping me to do so.

File: src/UI/Implementation/Component/Chart/ProgressMeter/FixedSize.php
Comment (klees) [inline src/UI/Implementation/Component/Chart/ProgressMeter/FixedSize.php]: Hi @qdittrich, sry for the long delay after @Amstutz last reply. I guess this part here is a leftover, since we removed the `width`-propery, right? Please remove it.

Still thanks for the cleanup, this looks really good now. And extra kudos for removing more lines than you have added. I'll merge after the leftover is removed.

**Review by klees — APPROVED** [review]

**Pull request merged**
